### PR TITLE
RONDB-589 Pushdown Aggregation - 24.10-main

### PR DESCRIPTION
**Pull request by KernelMaker**
Description:
1. API provides interfaces to write assembly-style code programs for aggregation queries.
2. API pushes the program with a scan request to RonDB data nodes.
3. Data nodes scan the rows from a specific table, for each row it gets, executing the pushed down program to generate the aggregation results locally.
4. when the scan is done on data nodes, they send the local result back to API.
5. API gathers and merges the results to get the final aggregation results.

Current abilities:
a.Supported column type:
  1. TINYINT
  2. TINYINT UNSIGNED
  3. SMALLINT
  4. SMALLINT UNSIGNED
  5. MEDIUMINT
  6. MEDIUMINT UNSIGNED
  7. INT
  8. INT UNSIGNED
  9. BIGINT
 10. BIGINT UNSIGNED
 11. FLOAT
 12. DOUBLE
 13. DECIMAL(Partially supported. Pushdown interpreter can load and parse DECIMAL column and try to convert it to BIGINT/BIGINT UNSIGNED/DOUBLE dynamically)

b.Supported arithmetic operator:
  1. Addition
  2. Subtraction
  3. Multiplication
  4. Division (DIV operator is supported. TODO: support / operator)
  5. Modulus

c. Support group by operation

d. Support working with pushdown filter

e. Support working with table-scan, index-scan( and range-scan).

### CODE REVIEW

File: storage/ndb/include/ndbapi/NdbAggregationCommon.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright [2024] <Copyright Hopsworks AB>
```
Comment (mronstro) [inline storage/ndb/include/ndbapi/NdbAggregationCommon.hpp]: Change this line in all files u changed to:
   Copyright (c) 2024, 2024, Hopsworks and/or its affiliates.

File: storage/ndb/src/kernel/blocks/dbtup/AggInterpreter.cpp [hunk 2]
```diff
@@ -0,0 +1,1572 @@
...
+        break;
+      case kOpMul:
+        reg_index = (value >> 12) & 0x0F;
+        reg_index2 = (value >> 8) & 0x0F;
+
+        ret = RegMulReg(registers_[reg_index], registers_[reg_index2],
+                  &registers_[reg_index]);
+        // assert(ret >= 0);
+        if (ret < 0) {
+          g_eventLogger->warning("Overflow[MUL], value is out of range");
```
Comment (mronstro) [inline storage/ndb/src/kernel/blocks/dbtup/AggInterpreter.cpp]: Change to g_eventLogger->debug, pretty sure warnings gets printed by default whereas debug only gets printed when specifically activated

Comment (mronstro) [reply storage/ndb/src/kernel/blocks/dbtup/AggInterpreter.cpp]: Good compromise between Alibaba style and RonDB style

File: storage/ndb/src/kernel/blocks/dbtup/DbtupExecQuery.cpp [hunk 3]
```diff
@@ -217,6 +219,59 @@ Uint32 Dbtup::copyAttrinfo(Uint32 storedProcId,
...
+        // Uint32 allocPageRef = 0;
+        // void* page_ptr = m_ctx.m_mm.alloc_page(RT_DBTUP_PAGE,
+        //                                 &allocPageRef,
+        //                                 Ndbd_mem_manager::NDB_ZONE_LE_30,
+        //                                 true);
+
+        void* page_ptr = lc_ndbd_pool_malloc(32 * 1024, RG_DATAMEM,
+                                             getThreadId(), false);
+        if (page_ptr == nullptr) {
+          g_eventLogger->error("Alloc mem for pushdown aggregation interpreter failed");
```
Comment (mronstro) [inline storage/ndb/src/kernel/blocks/dbtup/DbtupExecQuery.cpp:259]: ok use

**Review by mronstro — DISMISSED** [review]
Approved after fixing comments.
Question: Is it supported to use NdbQueryOperation as well, not sure I saw that support?
Great work.

**Comment by KernelMaker** [comment]
DONE

**Review by mronstro — APPROVED** [review]

**Pull request merged**